### PR TITLE
[WebNN EP] Support LSTM op

### PR DESCRIPTION
**Pull request by shiyi9801**

### CODE REVIEW

**Comment by shiyi9801** [comment]
@microsoft-github-policy-service agree company="Intel"

**Comment by shiyi9801** [comment]
@Honry, @fdwr, @guschmue PTAL, thanks!

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (shiyi9801) [reply onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:230]: Not sure if it's achievable. I tried to run the fallback path in cpu with `layout == 1`, and it reports error:
> onnxruntime::LSTMBase::LSTMBase(const OpKernelInfo &) layout_ == 0 was false. Batchwise recurrent operations (layout == 1) are not supported.

cc @fdwr for more insights, thanks!

**Comment by Honry** [comment]
@shiyi9801, pls. also update the https://github.com/microsoft/onnxruntime/blob/main/js/web/docs/webnn-operators.md.

**Comment by shiyi9801** [comment]
> @shiyi9801, pls. also update the https://github.com/microsoft/onnxruntime/blob/main/js/web/docs/webnn-operators.md.

@Honry Updated, please take a look, thanks :)

**Review by Honry — APPROVED** [review]
LGTM, thanks!

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (fdwr) [reply onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:187]: You may find this comment in the DML EP relevant:

https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRecurrentNeuralNetwork.cpp#L409-L422

```
    // layout=1 for batchwise operation is unsupported, added in opset 14 for RNN, GRU, and LSTM
    // (https://github.com/onnx/onnx/pull/3217, https://github.com/onnx/onnx/pull/2284).
    // Currently (2022-05-27) the ORT CPU execution provider (lstm_base.h) does not support it either,
    // with no models warranting it. When needed, it can be achieved with no new DML API's by just
    // swapping the size and strides in the TensorDesc before filling in the *_OPERATOR_DESC, where:
    //
    // layout=0: (default, consistent with opset 7)
    //      X.shape = [seq_length, batch_size, input_size]
    //      Y.shape = [seq_length, num_directions, batch_size, hidden_size]
    //      initial_h.shape = Y_h.shape = initial_c.shape = Y_c.shape = [num_directions, batch_size, hidden_size]
    // layout=1:
    //      X.shape = [batch_size, seq_length, input_size]
    //      Y.shape = [batch_size, seq_length, num_directions, hidden_size]
    //      initial_h.shape = Y_h.shape = initial_c.shape = Y_c.shape = [batch_size, num_directions, hidden_size]
```

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:45]: Same comment as on [GRU](https://github.com/microsoft/onnxruntime/pull/20405) - Should `!input_shape.empty()` be checked before deferencing index [0]? Can `GetShape` return true for a scalar, such that `input_shape[0]`?

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:49]: Should we check `input_defs.size() >= 3` before accessing `input_defs[2]`?

**Review by fdwr — DISMISSED** [review]
Minor comments, else LGTM.

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (Honry) [inline onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:51]: Missing `label` option for each WebNN operators, you can use node name as context.

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (Honry) [inline onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:193]: Missing `HasSupportedInputsImpl` check for inputs data type.

**Comment by fdwr** [comment]
New merge conflicts 🚧🚫.

**Comment by Honry** [comment]
@fdwr, @shiyi9801, how about hold this PR until the [`opSupportLimits`](https://github.com/microsoft/onnxruntime/pull/22025) is landed?

**Comment by shiyi9801** [comment]
> @fdwr, @shiyi9801, how about hold this PR until the [`opSupportLimits`](https://github.com/microsoft/onnxruntime/pull/22025) is landed?

agreed, will rebase after your PR is merged.

**Comment by fdwr** [comment]
> @fdwr, @shiyi9801, how about hold this PR until the [`opSupportLimits`](https://github.com/microsoft/onnxruntime/pull/22025) is landed?

That's complete now. 🏃‍♂️‍➡️

File: onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc
Comment (shiyi9801) [reply onnxruntime/core/providers/webnn/builders/impl/lstm_op_builder.cc:47]: The input is restricted to 3-D by ONNX, not sure if we need to add additional checks for that🤔

**Comment by shiyi9801** [comment]
@Honry @fdwr Rebased and fixed some bugs, please take another look, thanks!

**Review by fdwr — APPROVED** [review]
👍

**Comment by fdwr** [comment]
Retrying CI which timed out.

**Review by Honry — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
